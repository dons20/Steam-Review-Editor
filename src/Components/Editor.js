import React, { useState, useRef } from "react";
import { Editor as ReviewEditor } from "slate-react";
import { Value } from "slate";
import { isKeyHotkey } from "../Util/isHotkey";
import classes from "./editor.module.scss";
import Menu from "./Menu";

const databaseValue = JSON.parse(localStorage.getItem("content"));
const initialValue = {
    document: {
        nodes: [
            {
                object: "block",
                type: "paragraph",
                nodes: [
                    {
                        object: "text",
                        leaves: [
                            {
                                text: "Write your review here!"
                            }
                        ]
                    }
                ]
            }
        ]
    }
};
const starterValue = Value.fromJSON(databaseValue || initialValue);

/**
 * Define the default node type.
 *
 * @type {String}
 */

const DEFAULT_NODE = "paragraph";

/**
 * Define hotkey matchers.
 *
 * @type {Function}
 */

const isHeadingHotkey = isKeyHotkey("mod+h");
const isBoldHotkey = isKeyHotkey("mod+b");
const isUnderlinedHotkey = isKeyHotkey("mod+u");
const isItalicHotkey = isKeyHotkey("mod+i");
const isStrikethroughHotkey = isKeyHotkey("mod+-");
const isCodeHotkey = isKeyHotkey("mod+`");

/**
 * Deserialize the initial editor value.
 *
 * @type {JSX.Element}
 */

function Editor(props) {
    const [value, setValue] = useState(starterValue);
    const editor = useRef();

    /**
     * When a mark button is clicked, toggle the current mark.
     *
     * @param {Event} event
     * @param {String} type
     */

    const onClickMark = (event, type) => {
        event.preventDefault();
        editor.current.toggleMark(type);
    };

    /**
     * When a block button is clicked, toggle the block type.
     *
     * @param {Event} event
     * @param {String} type
     */

    const onClickBlock = (event, type) => {
        event.preventDefault();

        const { value } = editor.current;
        const { document } = value;

        // Handle everything but list buttons.
        if (type !== "unordered list" && type !== "ordered list") {
            const isActive = hasBlock(type);
            const isList = hasBlock("list-item");

            if (isList) {
                editor
                    .setBlocks(isActive ? DEFAULT_NODE : type)
                    .unwrapBlock("bulleted-list")
                    .unwrapBlock("numbered-list");
            } else {
                editor.current.setBlocks(isActive ? DEFAULT_NODE : type);
            }
        } else {
            // Handle the extra wrapping required for list buttons.
            const isList = hasBlock("list-item");
            const isType = value.blocks.some(block => {
                return !!document.getClosest(
                    block.key,
                    parent => parent.type === type
                );
            });

            if (isList && isType) {
                editor
                    .setBlocks(DEFAULT_NODE)
                    .unwrapBlock("bulleted-list")
                    .unwrapBlock("numbered-list");
            } else if (isList) {
                editor
                    .unwrapBlock(
                        type === "bulleted-list"
                            ? "numbered-list"
                            : "bulleted-list"
                    )
                    .wrapBlock(type);
            } else {
                editor.current.setBlocks("list-item").wrapBlock(type);
            }
        }
    };

    /**
     * Check if the current selection has a mark with `type` in it.
     *
     * @param {String} type
     * @return {Boolean}
     */

    const hasMark = type => {
        return value.activeMarks.some(mark => mark.type === type);
    };

    /**
     * Check if the any of the currently selected blocks are of `type`.
     *
     * @param {String} type
     * @return {Boolean}
     */

    const hasBlock = type => {
        return value.blocks.some(node => node.type === type);
    };

    /**
     * Render a Slate node.
     *
     * @param {Object} props
     * @return {Element}
     */

    const renderNode = (props, editor, next) => {
        const { attributes, children, node } = props;

        switch (node.type) {
            case "quote":
                return <blockquote {...attributes}>{children}</blockquote>;
            case "unordered list":
                return <ul {...attributes}>{children}</ul>;
            case "heading":
                return <h1 {...attributes}>{children}</h1>;
            case "list item":
                return <li {...attributes}>{children}</li>;
            case "ordered list":
                return <ol {...attributes}>{children}</ol>;
            default:
                return next();
        }
    };

    /**
     * Render a Slate mark.
     *
     * @param {Object} props
     * @return {Element}
     */

    const renderMark = (props, editor, next) => {
        const { children, mark, attributes } = props;

        switch (mark.type) {
            case "bold":
                return <strong {...attributes}>{children}</strong>;
            case "code":
                return <code {...attributes}>{children}</code>;
            case "italic":
                return <em {...attributes}>{children}</em>;
            case "underlined":
                return <u {...attributes}>{children}</u>;
            default:
                return next();
        }
    };

    /**
     * On change, save the new `value`.
     *
     * @param {Editor} editor
     */

    const onChange = ({ value }) => {
        setValue(value);
    };

    /**
     * On key down, if it's a formatting command toggle a mark.
     *
     * @param {Event} event
     * @param {Editor} editor
     * @return {Change}
     */

    const onKeyDown = (event, editor, next) => {
        let mark, node;

        if (isBoldHotkey(event)) {
            mark = "bold";
        } else if (isItalicHotkey(event)) {
            mark = "italic";
        } else if (isUnderlinedHotkey(event)) {
            mark = "underlined";
        } else if (isCodeHotkey(event)) {
            mark = "code";
        } else if (isHeadingHotkey(event)) {
            node = "heading";
        } else if (isStrikethroughHotkey(event)) {
            mark = "strikethrough";
        } else {
            return next();
        }

        event.preventDefault();
        if (mark) {
            editor.toggleMark(mark);
        } else {
            this.handleBlock(event, node);
        }
    };

    return (
        <div className={classes.root}>
            <div className={classes.menu}>
                <Menu
                    hasMark={hasMark}
                    hasBlock={hasBlock}
                    classes={classes}
                    onClickBlock={onClickBlock}
                    onClickMark={onClickMark}
                />
            </div>
            <div className={classes.editor}>
                <ReviewEditor
                    spellCheck
                    autoFocus
                    placeholder="Type your review here..."
                    ref={editor}
                    value={value}
                    onChange={onChange}
                    onKeyDown={onKeyDown}
                    renderNode={renderNode}
                    renderMark={renderMark}
                />
            </div>
        </div>
    );
}

export default Editor;
