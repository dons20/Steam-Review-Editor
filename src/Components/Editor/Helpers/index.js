import { insertLink, withImages, withLinks } from "./wrappers";
import Element from "./Element";
import Leaf from "./Leaf";
import { BlockButton, MarkButton, LinkButton } from "./Buttons";

export { insertLink, withImages, withLinks, Element, Leaf, BlockButton, MarkButton, LinkButton };
