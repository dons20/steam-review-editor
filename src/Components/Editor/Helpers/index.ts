import { insertLink, insertImage, insertTable, withImages, withLinks, withTables } from "./wrappers";
import {
  BlockButton,
  MarkButton,
  LinkButton,
  TableButton,
  QuoteButton,
  EraseButton,
  ImageButton,
  toggleBlock,
  toggleMark,
} from "./Buttons";
import Element from "./Element";
import Leaf from "./Leaf";

export {
  insertLink,
  insertImage,
  insertTable,
  withImages,
  withLinks,
  withTables,
  Element,
  Leaf,
  BlockButton,
  MarkButton,
  LinkButton,
  QuoteButton,
  TableButton,
  ImageButton,
  EraseButton,
  toggleBlock,
  toggleMark,
};
