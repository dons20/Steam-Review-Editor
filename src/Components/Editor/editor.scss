.editor__root {
    background-color: #fff;
    box-shadow: 0 2px 4px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.24);
    font-size: 18px;
    grid-row: 2;
    max-width: 1200px;
    height: 100%;
    width: 100%;

    &:focus {
        box-shadow: 0 3px 6px rgba(255, 255, 255, 0.5);
    }

    .code {
        background-color: rgb(32, 32, 32);
        color: #37d0ff;
        line-height: 2.5rem;
        padding: 6px 10px;
        position: relative;
        white-space: pre-wrap;
        &::before {
            color: #fff;
            content: "code";
            font-size: 11px;
            left: 5px;
            line-height: 10px;
            padding: 0;
            position: absolute;
            top: 1px;
        }
    }

    .custom {
        color: #888;
        cursor: pointer;
        flex: 1;
        text-transform: capitalize;
        text-align: center;
        transition: transform 0.1s ease-in;
        will-change: transform;
        &:hover {
            transform: translateY(-2px);
        }
        &[data-active] {
            color: #0063fd;
        }
    }

    .editor {
        min-height: 20vh;
        height: 100%;
        padding: 1.5rem;
    }

    .menu {
        align-items: center;
        border-bottom: 1px solid rgb(204, 204, 204);
        color: #555;
        display: grid;
        grid-template-columns: repeat(auto-fit, minmax(50px, 1fr));
        gap: 2rem 1rem;
        justify-content: space-around;
        padding: 2rem 0.5rem;
        user-select: none;
    }

    .noparse {
        background-color: #f3f3f3;
        line-height: 2.5rem;
        padding: 5px 10px;
        position: relative;
        white-space: pre-wrap;
        &::before {
            background-color: #fff;
            content: "no-parse";
            font-size: 11px;
            left: 5px;
            line-height: 0;
            padding: 0;
            position: absolute;
            top: 0;
        }
    }

    .spoiler {
        background-color: #222;
        color: #999;
        display: block;
        line-height: 2.5rem;
        padding: 5px 10px;
        position: relative;
        margin: 5px 0;
        &:hover {
            color: #fff;
        }
        &::before {
            background-color: #fff;
            color: #fff;
            content: "spoiler";
            font-size: 11px;
            left: 5px;
            line-height: 0;
            padding: 0;
            position: absolute;
            top: 10px;
        }
    }

    .table {
        border-collapse: collapse;
        margin: 15px 0;
        width: 100%;
        & td {
            border: 1px solid #696969;
            padding: 2px;
            text-align: center;
        }
    }

    .quote {
        background-color: #d1f3fc;
        border-left: 5px solid #26a5c5;
        color: #333333;
        font-style: italic;
        line-height: 2.5rem;
        margin: 5px 0;
        padding: 5px 30px;
        position: relative;
        white-space: pre-wrap;

        &::before {
            background-color: #fff;
            content: "\201C";
            color: #26a5c5;
            font-size: 2.5rem;
            font-family: Arial, Helvetica, sans-serif;
            left: 5px;
            line-height: 0;
            padding: 0;
            position: absolute;
            top: 20px;
        }
        &::after {
            color: #111;
            display: block;
            font-weight: bold;
        }
        &[data-author]::after {
            content: "- " attr(data-author);
        }
    }
}

@media only screen and (min-width: 769px) {
    .tooltip:before {
        right: inherit;
        text-align: center;
        width: initial;
        max-width: inherit;
        min-width: 100%;
    }
}
