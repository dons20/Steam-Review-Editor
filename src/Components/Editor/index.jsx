import React, { useEffect, useState, useContext, useMemo, useCallback } from "react";
import { Slate, Editable, withReact } from "slate-react";
import { isKeyHotkey } from "../../Util/isHotkey";
import initialJSONValue from "../value.json";
import { AppContext } from "../Content";
import Menu from "../Menu";

//eslint-disable-next-line
import { createEditor, Editor } from "slate";

import classes from "./editor.module.scss";

/** @type {JSON} */ const databaseValue = JSON.parse(localStorage.getItem("content"));
/** @type {Object} */ const starterValue = databaseValue || initialJSONValue;

/** @type {String} */ const DEFAULT_NODE = "paragraph";

/** @type {(value) => Boolean} */ const isHeadingHotkey = isKeyHotkey("mod+h");
/** @type {(value) => Boolean} */ const isBoldHotkey = isKeyHotkey("mod+b");
/** @type {(value) => Boolean} */ const isUnderlinedHotkey = isKeyHotkey("mod+u");
/** @type {(value) => Boolean} */ const isItalicHotkey = isKeyHotkey("mod+i");
/** @type {(value) => Boolean} */ const isStrikethroughHotkey = isKeyHotkey("mod+-");
/** @type {(value) => Boolean} */ const isCodeHotkey = isKeyHotkey("mod+`");
/** @type {(value) => Boolean} */ const isNewlineHotKey = isKeyHotkey("shift+enter");
/** @type {(value) => Boolean} */ const isDeleteHotKey = isKeyHotkey("delete");
/** @type {(value) => Boolean} */ const isSelectAllHotKey = isKeyHotkey("mod+a");

/**
 * A change helper to standardize wrapping links.
 *
 * @param {Editor} editor
 * @param {String} href
 */

function wrapLink(editor, href) {
    editor.wrapInline({
        type: "link",
        data: { href }
    });

    editor.moveToEnd();
}

/**
 * A change helper to standardize unwrapping links.
 *
 * @param {Editor} editor
 */

function unwrapLink(editor) {
    editor.unwrapInline("link");
}

/**
 * Hook to trigger a timeout
 *
 * @param {Function} closeTimer
 * @param {Function} saveEditor
 * @param {Object} value
 * @param {Number} delay
 * @param {Boolean} active
 */
function useTimeout(closeTimer, saveEditor, value, delay, active) {
    // Set up the interval.
    useEffect(() => {
        function save() {
            saveEditor(value);
            closeTimer();
        }
        if (delay !== null && active) {
            let id = setTimeout(save, delay);
            return () => clearInterval(id);
        }
    }, [active, delay, closeTimer, saveEditor, value]);
}

/**
 * Defines the core editor component
 *
 * @returns {JSX.Element}
 */

function ReviewEditor() {
    const editor = useMemo(() => withReact(createEditor()), []);

    /** @type {[Object, React.SetStateAction<any>]} */
    const [value, setValue] = useState(starterValue);

    /** @type {[Boolean, React.SetStateAction<any>]} */
    const [timerActive, setTimerActive] = useState(true);

    /** @type {[Boolean, React.SetStateAction<any>]} */
    const [isAllSelected, shouldSelectAll] = useState(false);

    /** @type {Function} */
    const setPreviewContent = useContext(AppContext);

    /**
     * Adds event listener to save editor content before refreshes/navigation changes
     */
    useEffect(() => {
        function runBeforeExit() {
            saveEditor(value);
        }

        window.addEventListener("beforeunload", runBeforeExit);
        return function cleanup() {
            window.removeEventListener("beforeunload", runBeforeExit);
        };
    });

    /**
     * Saves the content of the editor to localStorage
     */
    const saveEditor = () => {
        //let val = value.toJSON();
        const content = JSON.stringify(value);
        localStorage.setItem("content", content);
        setPreviewContent(value);
        setTimerActive(false);
    };

    /**
     * Auto-saves editor content after delay.
     */
    useTimeout(() => setTimerActive(false), saveEditor, value, 1500, timerActive);

    /**
     * Check if the current selection has a mark with `type` in it.
     *
     * @param {String} type
     * @return {Boolean}
     */

    const hasMark = type => {
        if (editor.current) {
            const { value } = editor.current;
            return value.activeMarks.some(mark => mark.type === type);
        }
    };

    /**
     * Check if the any of the currently selected blocks are of `type`.
     *
     * @param {String} type
     * @return {Boolean}
     */

    const hasBlock = type => {
        if (editor.current) {
            const { value } = editor.current;
            return value.blocks.some(node => node.type === type);
        }
    };

    /**
     * Check whether the current selection has a link in it.
     *
     * @return {Boolean} hasLinks
     */

    const hasLinks = () => {
        if (editor.current) {
            const { value } = editor.current;
            return value.inlines.some(inline => inline.type === "link");
        }
    };

    /**
     * When a mark button is clicked, toggle the current mark.
     *
     * @param {Event} event
     * @param {String} type
     */

    const onClickMark = (event, type) => {
        event.preventDefault();
        editor.current.toggleMark(type);
        //Small hack to allow correct editor reference to be focused
        setTimeout(() => {
            editor.current.focus();
        });
    };

    /**
     * When a block button is clicked, toggle the block type.
     *
     * @param {React.KeyboardEvent} event
     * @param {String} type
     */

    const onClickBlock = (event, type) => {
        event.preventDefault();

        const { value } = editor.current;
        const { document } = value;
        // Handle list blocks
        if (type === "unordered list" || type === "ordered list") {
            // Handle the extra wrapping required for list buttons.
            const isList = hasBlock("list item");
            const isType = value.blocks.some(block => {
                return !!document.getClosest(block.key, parent => parent.type === type);
            });

            if (isList && isType) {
                editor.current
                    .setBlocks(DEFAULT_NODE)
                    .unwrapBlock("unordered list")
                    .unwrapBlock("ordered list");
            } else if (isList) {
                editor.current
                    .unwrapBlock(type === "unordered list" ? "ordered list" : "unordered list")
                    .wrapBlock(type);
            } else {
                editor.current.setBlocks("list item").wrapBlock(type);
            }
        } else if (type === "table") {
            //Handle table blocks
            const isTable = editor.current.isSelectionInTable();
            if (isTable) {
                editor.current.removeTable();
            } else {
                let rows = parseInt(window.prompt("Enter the number of rows:"));
                if (!rows) return;
                let columns = parseInt(window.prompt("Enter the number of columns:"));
                if (!columns) return;
                editor.current.insertTable(columns, rows);
            }
        } else {
            //Handle every other block
            const isActive = hasBlock(type);
            const isList = hasBlock("list item");
            let author = null,
                img = null;

            if (type === "quote" && !isActive) {
                author = window.prompt("(Optional) Enter the author name:");
            } else if (type === "image" && !isActive) {
                img = window.prompt("Please enter the image url:");
                if (!img) return;
            }

            if (isList) {
                editor.current
                    .setBlocks({ type: isActive ? DEFAULT_NODE : type })
                    .unwrapBlock("unordered list")
                    .unwrapBlock("ordered list");
                if (author) editor.current.setBlocks({ data: { author } });
                if (img) editor.current.setBlocks({ data: { img } });
            } else {
                editor.current.setBlocks({
                    type: isActive ? DEFAULT_NODE : type
                });
                if (author) editor.current.setBlocks({ data: { author } });
                if (img) editor.current.setBlocks({ data: { img } });
            }
        }

        //Small hack to allow correct editor reference to be focused
        setTimeout(() => {
            editor.current.focus();
        });
    };

    /**
     * Performs Custom behaviour on click
     *
     * @param {MouseEvent} event
     * @param {String} type
     */
    const onClickCustom = (event, type) => {
        event.preventDefault();

        /** @type {Editor}*/
        const { value: currentValue } = editor.current;

        switch (type) {
            case "reset to default":
                //Select all and delete
                if (currentValue.document.nodes.size > 0) {
                    editor.current.moveToRangeOfDocument().delete();
                } else {
                    let willReset = window.confirm(
                        "The editor has no child nodes. Will you reset to a default value? (Will remove previous history)"
                    );
                    if (willReset) setValue(initialJSONValue);
                }
                break;
            case "link":
                const editorHasLinks = hasLinks();

                if (editorHasLinks) {
                    editor.current.command(unwrapLink);
                } else if (currentValue.selection.isExpanded) {
                    const href = window.prompt("Enter the URL of the link:");

                    if (href === null) {
                        return;
                    }

                    editor.current.command(wrapLink, href);
                } else {
                    const href = window.prompt("Enter the URL of the link:");

                    if (href === null) {
                        return;
                    }

                    const text = window.prompt("Enter the text for the link:");

                    if (text === null) {
                        return;
                    }

                    editor.current
                        .insertText(text)
                        .moveFocusBackward(text.length)
                        .command(wrapLink, href);
                }
                break;
            default:
                break;
        }
        //Small hack to allow correct editor reference to be focused
        setTimeout(() => {
            editor.current.focus();
        });
    };

    /**
     * Render a Slate node.
     *
     * @param {Object} props
     * @param {ReviewEditor} editor
     * @return {JSX.Element}
     */

    const renderElement = useCallback(props => {
        const { attributes, children, element } = props;

        switch (element.type) {
            case "heading":
                return <h1 {...attributes}>{children}</h1>;
            case "quote":
                const { data } = element;
                const author = data.get("author");
                return (
                    <blockquote className={classes.quote} data-author={author} {...attributes}>
                        {children}
                    </blockquote>
                );
            case "unordered list":
                return <ul {...attributes}>{children}</ul>;
            case "ordered list":
                return <ol {...attributes}>{children}</ol>;
            case "list item":
                return <li {...attributes}>{children}</li>;
            case "code":
                return (
                    <pre className={classes.code} {...attributes}>
                        {children}
                    </pre>
                );
            case "spoiler":
                return (
                    <div className={classes.spoiler} {...attributes}>
                        {children}
                    </div>
                );
            case "noparse":
                return (
                    <pre className={classes.noparse} {...attributes}>
                        {children}
                    </pre>
                );
            case "table":
                return (
                    <table className={classes.table}>
                        <tbody {...attributes}>{children}</tbody>
                    </table>
                );
            case "table-row":
                return <tr {...attributes}>{children}</tr>;
            case "table-cell":
                return <td {...attributes}>{children}</td>;
            case "link": {
                const { data } = element;
                const href = data.get("href");
                return (
                    <a {...attributes} href={href}>
                        {children}
                    </a>
                );
            }
            case "image": {
                const { data } = element;
                const img = data.get("img");
                return (
                    <div {...attributes}>
                        <img src={img} alt={"Image Link: " + img} />
                        {children}
                    </div>
                );
            }
            default:
                return <p {...attributes}>{children}</p>;
        }
    }, []);

    /**
     * Render a Slate leaf.
     *
     * @param {Object} props
     * @return {JSX.Element}
     */

    const renderLeaf = useCallback(props => {
        const { children, leaf, attributes } = props;

        if (leaf.bold) return <strong {...attributes}>{children}</strong>;
        if (leaf.italic) return <em {...attributes}>{children}</em>;
        if (leaf.underlined) return <u {...attributes}>{children}</u>;
        if (leaf.strikethrough) return <s {...attributes}>{children}</s>;
        if (leaf.image) {
            return (
                <img alt="" {...attributes}>
                    {children}
                </img>
            );
        }
        return <span {...attributes}>{children}</span>;
    }, []);

    /**
     * On change, save the new `value`.
     */

    const onChange = newValue => {
        // Check to see if the document has changed before saving.
        if (value.document !== newValue.document) {
            setTimerActive(true);
        }

        setValue(newValue);
    };

    /**
     * On key down, if it's a formatting command toggle a mark.
     *
     * @param {React.KeyboardEvent} event
     * @param {Editor} editor
     * @param {() => any} next
     */

    const onKeyDown = (event, editor, next) => {
        let mark, node;

        if (isBoldHotkey(event)) {
            mark = "bold";
        } else if (isItalicHotkey(event)) {
            mark = "italic";
        } else if (isUnderlinedHotkey(event)) {
            mark = "underlined";
        } else if (isCodeHotkey(event)) {
            mark = "code";
        } else if (isHeadingHotkey(event)) {
            node = "heading";
        } else if (isStrikethroughHotkey(event)) {
            mark = "strikethrough";
        } else if (isNewlineHotKey(event)) {
            return editor.insertText("\n");
        } else if (isSelectAllHotKey(event)) {
            shouldSelectAll(!isAllSelected);
            return next();
        } else if (isDeleteHotKey(event)) {
            if (isAllSelected) {
                editor.setBlocks(DEFAULT_NODE);
                shouldSelectAll(false);
                return next();
            } else {
                return next();
            }
        } else {
            if (isAllSelected) shouldSelectAll(false);
            return next();
        }

        event.preventDefault();
        if (mark) {
            editor.toggleMark(mark);
        } else {
            onClickBlock(event, node);
        }
    };

    const onSelect = (event, editor, next) => {
        if (isAllSelected) shouldSelectAll(false);
        return next();
    };

    return (
        <div className={classes.root}>
            <div className={classes.menu}>
                <Menu
                    editor={editor}
                    hasMark={hasMark}
                    hasBlock={hasBlock}
                    hasLinks={hasLinks}
                    classes={classes}
                    onClickMark={onClickMark}
                    onClickBlock={onClickBlock}
                    onClickCustom={onClickCustom}
                />
            </div>
            <div className={classes.editor}>
                <Slate
                    editor={editor}
                    value={value}
                    onChange={onChange}
                    //spellCheck
                    //autoFocus
                    //placeholder="Type your review here..."
                    //defaultValue={value}
                    //onSelect={onSelect}
                >
                    <Editable onKeyDown={onKeyDown} renderElement={renderElement} renderLeaf={renderLeaf} />
                </Slate>
            </div>
        </div>
    );
}

export default ReviewEditor;
