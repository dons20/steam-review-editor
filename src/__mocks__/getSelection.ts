window.getSelection = jest.fn().mockImplementation(() => {
  return {
    anchorNode: Node,
    anchorOffset: Number,
    focusNode: Node,
    focusOffset: Number,
    isCollapsed: true,
    rangeCount: 0,
    type: null,
  };
});
