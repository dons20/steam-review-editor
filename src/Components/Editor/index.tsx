import React, { useEffect, useState, useContext, useCallback, useRef } from "react";
import { createReactEditorJS } from "react-editor-js";
import { EDITOR_JS_TOOLS } from "./Helpers/tools";
import { AppContext } from "Components/Content";

import "./editor.scss";

const EditorJs = createReactEditorJS();

/** @type {JSON} */ const databaseValue = JSON.parse(localStorage.getItem("content"));
/** @type {Object} */ const starterValue = databaseValue || undefined;

/**
 * Hook to trigger a timeout
 *
 * @param {Function} closeTimer
 * @param {Function} saveEditor
 * @param {Object} value
 * @param {Number} delay
 * @param {Boolean} active
 */
function useTimeout(closeTimer, saveEditor, value, delay, active) {
  // Set up the interval.
  useEffect(() => {
    function save() {
      saveEditor(value);
      closeTimer();
    }
    if (delay !== null && active) {
      let id = setTimeout(save, delay);
      return () => clearInterval(id);
    }
  }, [active, delay, closeTimer, saveEditor, value]);
}

/**
 * Defines the core editor component
 *
 * @returns {JSX.Element}
 */

function ReviewEditor() {
  const editorRef = useRef(null);

  /** @type {[Object, React.Dispatch<React.SetStateAction<any>>]} */
  const [data, setData] = useState(starterValue);

  /** @type {[Boolean, React.Dispatch<React.SetStateAction<any>>]} */
  const [timerActive, setTimerActive] = useState(true);

  /** @type {{ setHTMLContent: React.Dispatch<React.SetStateAction<any>>, notify: Function }} */
  const { setHTMLContent, notify } = useContext(AppContext);

  const currentValue = useRef(data);

  /**
   * Saves the content of the editor to localStorage
   */
  const saveEditor = useCallback(async () => {
    const savedData = await editorRef.current.save();
    const content = JSON.stringify(savedData);
    localStorage.setItem("content", content);
    //setHTMLContent(data);
    console.log(savedData);
    setTimerActive(false);
    notify("Latest changes saved âœ…");
    currentValue.current = data;
  }, [data, notify]);

  /**
   * Adds event listener to save editor content before refreshes/navigation changes
   */
  useEffect(() => {
    function runBeforeExit() {
      saveEditor();
    }

    //setHTMLContent(data);

    window.addEventListener("beforeunload", runBeforeExit);

    return function cleanup() {
      window.removeEventListener("beforeunload", runBeforeExit);
    };
  }, [saveEditor, setHTMLContent, data]);

  /**
   * Auto-saves editor content after delay.
   */
  useTimeout(() => setTimerActive(false), saveEditor, data, 10000, timerActive);

  /**
   * On change, save the new `value`.
   */

  const onChange = newValue => {
    // Check to see if the value has changed before saving.
    if (!Object.is(currentValue.current, newValue)) {
      setTimerActive(true);
    }

    setData(newValue);
  };

  const handleInitialize = React.useCallback(instance => {
    editorRef.current = instance;
  }, []);

  //   const handleSave = React.useCallback(async () => {
  // 	const savedData = await editorRef.current.save();
  //   }, []);

  return (
    <div className="editor__root">
      <EditorJs
        // data={data}
        placeholder="Type your review here..."
        // instanceRef={instance => (editorRef.current = instance)}
        onInitialize={handleInitialize}
        onChange={onChange}
        autofocus={true}
        tools={EDITOR_JS_TOOLS}
      />
    </div>
  );
}

export default ReviewEditor;
