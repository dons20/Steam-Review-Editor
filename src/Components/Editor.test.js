import "./__mocks__/getSelection";

import React from "react";
import ReactDOM from "react-dom";
import { Value } from "slate";

import Editor from "./Editor";
import EditorValue from "./testValue.json";
import SteamMarkup from "../Util/slate-steam-serializer";

it("renders the editor", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Editor />, div);
    ReactDOM.unmountComponentAtNode(div);
});

it("converts an editor value to steam markup", () => {
    const toSteam = new SteamMarkup({
        rules: [
            {
                /**
                 * @param {import('slate').Document} obj
                 * @param {Array} children
                 */
                serialize(obj, children) {
                    if (obj.object === "block") {
                        switch (obj.type) {
                            case "heading":
                                return "[h1]" + children + "[/h1]";
                        }
                    }
                }
            }
        ]
    });
    const newVal = toSteam.serialize(Value.fromJSON(EditorValue));
    console.log(newVal);
    expect(newVal).toMatchSnapshot();
});
