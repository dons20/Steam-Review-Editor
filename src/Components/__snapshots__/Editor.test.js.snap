// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`converts an editor value to steam markup 1`] = `
SteamMarkup {
  "cruftNewline": [Function],
  "defaultBlock": Object {
    "type": "paragraph",
  },
  "parseMarkup": [Function],
  "rules": Array [
    Object {
      "serialize": [Function],
    },
    Object {
      "deserialize": [Function],
      "serialize": [Function],
    },
  ],
  "serialize": [Function],
  "serializeLeaf": [Function],
  "serializeNode": [Function],
  "serializeString": [Function],
}
`;
